### ISSUE TITLE
Manually specify OIDC configuration?

### ISSUE THREAD

**Issue by agilenut** [IdentityModel/IdentityModel.AspNetCore#208]
I have an application that is composed of multiple micro front-ends that are on the same domain and share an authentication cookie. The primary shell is responsible for the OpenID Connect authentication and the rest just use the cookie. But I'd like to use this library to keep the cookie up to date with refreshed access tokens. Currently, the library fails with a null reference exception. I believe this is because it is running in a context that does not have the OpenID Connect authentication scheme setup. Is there an easy way to manually specify the configuration to bypass the automatic lookup?

**Comment by VictorioBerra** [comment]
I dont fully follow. Is the micro frontends just a single SPA on a single domain? You make it sound like they are in iframes?

This library is for the server side, if you do backend-for-frontend pattern you dont need to worry about tokens or anything. Just do cookie auth as usual and use the client access stuff in this library.

Alternatively, you can easily swap out a bunch of the services if you need ANYTHING custom, its very easy. I highly recommend you read the source, theres not much of it.

**Comment by agilenut** [comment]
Sorry, “micro front-ends” might be a bit misleading. Each app is a blazer server-side app that implements a vertical slice of the system so that it can be broken up into smaller composable parts and worked on by different teams. 

The plan is for the “primary app” to control the oidc flow to get the access and refresh tokens. Then create a cookie that each “sub app” (on the same domain) can use to identify the user and, if needed, refresh the token. 

I had actually already written an extension to the cookie auth handler on principal validated event to refresh the access token when it is near expiration. But then I came across the package and thought it would be better than doing it myself. 

I looked at the code and it seems to be getting the url info from oidc options. Those aren’t configured in the sub apps. Based on the properties of this options class it seems to be populated after the discovery endpoint call. I probably need to look a bit more at what actually configures those options and work my way back from that. I was just hoping to not have to register the full oidc middleware in each sub app if it isn’t actually used.

**Comment by VictorioBerra** [comment]
I think im understanding a little better now. So, you would pull this library into each Blazor Server app, and then they all read and decrypt the cookie by using a shared key. And each sub-app makes API calls as usual and the token management httpclient middleware would refresh it if need be, and store it locally in its singleton cache.

I am a bit confused about how the primary app can be in charge of the initial OIDC flow. If a sub-app opens the cookie, and finds the access token expired, it would have to refresh it. In order to do that, it would need a client ID and have a proper redirect URI and all that.

Alternatively, I wonder if you could just swap out the [cache ](https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/27e966bf67/src/AccessTokenManagement/ClientAccessToken/ClientAccessTokenCache.cs) for a redis based one? This way every app could refresh the cache that is shared by all apps?

Sorry if im missing the point here, still trying to form a mental picture of the sub-apps and the layout.

**Comment by agilenut** [comment]
Yes. You've got it. 

Regarding the OIDC flow, the thought was that the primary app would initiate redirect to the authorize endpoint. If a user requests a path in a sub-app, it would first redirect to the primary app, then the primary app would redirect back. But, you are correct. Each sub-app would need to handle the back-channel call to refresh the token itself which is why I'm looking at this package. So, it does need to have the client ID and client secret. I don't think it should need the redirect URL since the token call would be back-channel. 

Regarding cache, yes, I considered a distributed cache. We may do that at some point but, we might start with keeping it in the cookie. Though the cookie size is already large enough that it creates 2 chunks.

**Comment by leastprivilege** [comment]
I probably did not design the library for exactly what you are trying to achieve ;) So feel free to do your own investigation and let me know what you found out...

**Comment by agilenut** [comment]
Understood :-)

From what I've tested today, the simplest way to make this work is to go ahead and register the OIDC auth scheme and handler along with the options so that it is there when this package looks for them. To prevent the OIDC scheme from running, I just set the default security scheme to cookies:

`services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)`

and then I configure the options for this package with the OIDC scheme so that it knows the scheme is there even though it is not the default challenge scheme.

`services.AddAccessTokenManagement(options => options.User.Scheme = OktaDefaults.MvcAuthenticationScheme);`

This seems to work with my initial testing. The main downside is I have to register the OIDC scheme and handler which isn't actually being used. But I'm not sure that downside is worth rolling my own method for registering the options and calling the post configure options that MS does.
